kaitaistruct.py

```diff
@@ -317,6 +386,9 @@ def read_bytes_term(self, term, include_term, consume_term, eos_error):
             else:
                 r += c
 
+    def write_bytes_term(self, data, term=b'\0'):
```

Comment:
AFAIR, no such method is used by the compiler branch [`serialization`](https://github.com/kaitai-io/kaitai_struct_compiler/tree/serialization) currently (it writes the terminator byte manually using `io.writeU1()` in Java if I remember correctly, so it doesn't make sense to have it in the runtime library. Moreover, using `terminator` as padding is actually incorrect, because `terminator` is not the same as `pad-right`. Terminator is a single byte signaling the end of some particular data, whereas padding is a continuous sequence of the `pad-right` byte after the end of data to fill the space reserved for the field using `size` or `size-eos: true` in the `.ksy` spec.

The `terminator` and `pad-right` bytes are _usually_ the same, but they don't have to be, see [`bytes_pad_term. ksy`](https://github.com/kaitai-io/kaitai_struct_tests/blob/master/formats/bytes_pad_term.ksy).

Reply:
There is currently **no** Python serialization implementation whatsoever. So this entire PR is more of a suggestion and, currently, manual implementation helper. There is no need to do exactly the same thing Java does because Python's a different language and serialization in different languages can differ so it needn't be copied from one another as for ease of maintaining together. 

The main idea of separate function for terminator is possibility of 2 and 4 byte string support where we can't actually write a single byte terminator. And if the compuller would write it all directly just what do we even need the runtime for.

Write padding should be done using `write_bytes` and termination with `write_bytes_term` for code readability purposes. So if the compiller will write `write_bytes_term` it'd be easier to understand that it's a terminated block and not some random byte written. So there is no contradiction between `terminator` and `pad-right`.

Reply:
> So this entire PR is more of a suggestion and, currently, **manual implementation helper**.

Kaitai Struct runtime libraries are not primarily intended to be a _manual implementation helper_. There are probably tons of other Python libraries that would serve this purpose better, or you can write your own. I'm sorry, but I'll not accept this as an argument.

> There is no need to do exactly the same thing Java does because Python's a different language

I'd say the opposite: there **must** be a need why do another thing than the existing languages because of robustness and maintainability. Please prepare a bulletproof explanation preferably with a solid reproducible benchmark (if it's from some performance reasons) every time you suggest such a change.

A few months ago I wanted to add support for [bit-sized integers](http://doc.kaitai.io/user_guide.html#_bit_sized_integers) in little-endian byte order (only big-endian bit ints had been supported before), which I implemented by renaming the existing `read_bits_int()` method to `read_bits_int_be()` and adding a new method `read_bits_int_le()` as its counterpart (which is the same algorithm with 2 small changes). It went quite well, because most runtime libraries were using the exact same logic and approach, so I quickly saw any differences which could cause bugs and could fix them immediately. In fact, when thinking about the common algorithm for a while, I was able to figure out a tiny optimization, which I applied to the rest of runtimes. 

The `ReadBitsInt()` method in Go was the only one implemented in a **completely** different way. There was no sign of the common algorithm I was used to, and thus I had to spend several hours figuring out how it works so that I could add the `ReadBitsIntLe()` counterpart. In the end, I gave up and rewrote the method completely from scratch using the established algorithm, which took me only a few minutes (see https://github.com/kaitai-io/kaitai_struct_go_runtime/pull/25). 

It's not only about maintainability, but the uncommon `ReadBitsInt()` algorithm actually had some real bugs that nobody noticed for a long time (https://github.com/kaitai-io/kaitai_struct_go_runtime/pull/22). I'd say unsurprisingly, because only a small subset of people are interested in Go codegen, and only units of people probably understood that custom algorithm.

The reason that _Python is a different language of Java_ is completely out of the game. You know very well that you can implement the same algorithm in all programming languages using often only simple 1:1 conversion: if you assign to a variable `myVar` in Java, in Python you assign to a variable `my_var`, and that's the whole difference between them.

> The main idea of separate function for terminator is possibility of 2 and 4 byte string support where we can't actually write a single byte terminator.

No, don't think about adding serialization support for things we don't have even parsing support, without **any** previous analysis of the implementation approach you choose. First read https://github.com/kaitai-io/kaitai_struct/issues/158 and https://github.com/kaitai-io/kaitai_struct/issues/187, then suggest some solution, which can be further discussed.

> Write padding should be done using `write_bytes` and termination with `write_bytes_term` for code readability purposes. So if the compiler will write `write_bytes_term` it'd be easier to understand that it's a terminated block and not some random bytes written.

This doesn't seem to me like a solid argument to me. It's a matter of personal preference, one could say that it's more explicit and transparent when you write the terminator manually. There has to be some other reason why deviate from the established convention from Java.

Reply:
> Kaitai Struct runtime libraries are not primarily intended to be a _manual implementation helper_
This is not _manual implementation helper_ **per se** it is one **until** it's implemented on the compiler side. And it should be at some point because there is a serious need for native serialization support.

>There has to be some other reason why deviate from the established convention from Java.
There is no _convention_ from a single Java branch as it is in development and the only implementation. So anything should work.

Anyway it's a chicken and egg problem here. We need to write compiler support before runtime or runtime before compiler support. I decided to start from the runtime. 

Padding and termination can be done this way properly. That's because termination is a case of padding-right with a (single in case of ASCII C-strings) termination byte and it makes your whole first point invalid.